AutoRepair documentation improvements (#22)
jaydeepkumar1984/cassandra: doc/modules/cassandra/pages/auto-repair/overview.adoc

= Overview of Auto Repair
:navtitle: Auto Repair overview
:description: Auto Repair concepts - How it works, how to configure it, and more.
:keywords: CEP-37

> Repair orchestration inside Apache Cassandra. This fully Automated Repair scheduler inside Cassandra does not depend on the control plane, significantly reducing our operational overhead.

Reviewer: Just a suggestion to make this more clear for users:

```suggestion
Auto Repair is a fully automated scheduler that provides repair orchestration inside of Apache Cassandra.

This significantly reduces operational overhead by removing the need for operators to deploy tooling to submit and manage repairs.

```

Author: Done

> At a high level, a dedicated thread pool is assigned to the repair scheduler. The repair scheduler inside Cassandra maintains a new replicated table under a distributed system_distributed keyspace. This table maintains the repair history for all the nodes, such as when it was repaired the last time, etc. The scheduler will pick the node(s) that run the repair first and continue orchestration to ensure Every table and all of their token ranges are repaired. The algorithm can also run repairs simultaneously on multiple nodes and splits the token range into subranges with the necessary retry to handle transient failures. The automatic repair runs as soon as we start a Cassandra cluster, like Compaction, and does not require human-in-the-loop.

Reviewer: a new replicated table under a distributed system_distributed keyspace.

to:

a new replicated table, system_distributed.auto_repair_history.

Reviewer: remove ', etc.'.

Reviewer: Every table -> every table.

Compaction -> compaction.

Author: Done

The scheduler supports Full, Incremental, and Preview repair types today with the following features. A new repair type, such as Paxos repair and any future repair types, should be extended with minimal dev work!

=== Features
- Capability to run on multiple nodes simultaneously
- Default implementation and an interface to override the data set being repaired per repair session
- Ability to extend the token split algorithms with the following two implementations readily available:

> - Splits token ranges by putting a max cap on the size of data being repaired as part of one repair session as well as a max cap at a table level - this is crucial for Incremental Repair (Default)

Reviewer: ```suggestion
- Splits token ranges by putting a cap on the size of data being repaired as part of one repair session as well as a max cap at a per-schedule level (Default)
```

Reviewer: should also indent this (and the one below) as they go under the 'Ability' one above.

Author: Done

- Split tokens evenly based on the number of splits specified
- A new CQL table property to do the following:

> - Disable repair type at a table level if one wants the scheduler to skip one or more tables

Reviewer: indent these two

Author: Done

- Setting repair priority for certain tables to prioritize those tables over others
- Enabling/Disabling scheduler for each repair type dynamically
- Per Data Center per job configuration
- Rich Configuration for each repair type, e.g., full, incremental, or preview_repair
- Enough observability so an operator can configure alarms depending on their need

=== Yaml Configuration
[cols=",,",options="header",]
|===
| Name | Default | Description
| enabled | false | Enable/Disable auto repair scheduler

> | min_repair_interval | 24h | Minimum time in hours between repairing the same node again. This is useful for extremely tiny clusters, say 5 nodes, which finishes. repair quicly. The default is 24 hours. This means that if the scheduler finishes one round on all the nodes in < 24 hours. On a given node it won’t start a new repair round until the last repair conducted on a given node is < 24 hours.

Reviewer: need to break out the `global_settings`/repair level settings from the top level ones.

Author: Done
